### DISCUSSION TITLE
Can't reset nested object value to empty object, resets deepest object only. How to reset completely?

### DISCUSSION THREAD

**Discussion by AveryTheDev** [react-hook-form/react-hook-form#7644] [Q&A] [answered]
If I have a form set up so that its values assigned look like so 

```js
 const formData = {
    isValid: {true: false},
    tag: {foo: true, bar: true}
 }
```

And that its default values are the following

```js
 const defaultValues = {
   isValid: undefined,
   tag: undefined
 }
```

When I attempt to reset the form values using `reset(defaultValues)`, I get the following instead of the default values I expected.

```js
console.log(formData); 

/*
{
    isValid: {true: undefined},
    tag: {foo: undefined, bar: undefined}
 }
*/
```

I wanted my nested object values to be completely reset to whatever the default value is. Instead I see my nested object's keys still present with only their values changed. 

How can I get the `formData` to look like `defaultValue`'s? I tried using `setValue` and assigning the nested objects a value directly but I couldn't get it to work. Not certain if my syntax is wrong.

```js
 const resetAllValues = () => {
    reset(defaultValues);
    setValue('tag', {}); //only the rest value is seen
   console.log(formData); 

/*
  {
    isValid: {true: undefined},
    tag: {foo: undefined, bar: undefined}
   }
*/
};
```

Thoughts?

**Comment by msageryd** [comment]
I cannot offer an answer. It seems quite odd that your "undefined" gets assigned deep into your object.

But I do know that `undefined` is not supported in RHF, as per information from @bluebill1049. I'm having trouble with this myself. To add to the confusion, `null` is not supported by input fields on the web (a React thing), but it seems to be supported on iOS and Android using React-Native. The way to reset fields on the web seems to be to assign empty strings as default values. I'm resetting by assigning null i React Native.

**Comment by Moshyfawn** [comment]
If you look at [this CSB](https://codesandbox.io/s/xenodochial-yonath-1v9qp?file=/src/App.tsx), you can see that your actual default form value is
```js
{}
```
given that you provide
```js
{
  isValid: undefined,
  tag: undefined
}
```
So, if you try to fully reset your form via [the`reset` from method](https://react-hook-form.com/api/useform/reset/#main), it will rightfully reset it to
```js
{}
```

**Reply by msageryd** [reply]
If you change `undefined` to `null` in the CSB the form resets.
```
{
  isValid: null,
  tag: null
}
```

**Reply by Moshyfawn** [reply]
> If you change `undefined` to `null` in the CSB the form resets.
> 
> ```
> {
>   isValid: null,
>   tag: null
> }
> ```

Yep, 'cuz in JS `undefined` is an initialized variable without an assigned value and `null` is basically a variable initialized with an empty value.

**Reply by msageryd** [reply]
Both the checkboxes and the formData gets reset to null. If these where TextInputs, they'd need to be reset with empty strings.

Is `null` the correct way to reset checkbox fields then?

**Reply by AveryTheDev** [reply]
Please see https://github.com/react-hook-form/react-hook-form/discussions/7644#discussioncomment-2036017 for that does not work

**Comment by AveryTheDev** [comment]
When I changed the value to `null` I got 

```
/*
  {
    isValid: {true: null},
    tag: {foo: null, bar: null}
   }
*/
```

instead of

```js
{
  isValid: null,
  tag: null
}
```

If I attempt to reset default values it as an empty object, nothing changes. Assigning an empty object to the keys that I want to use result in.

```
/*
  {
    isValid: {true:{}},
    tag: {foo:{}, bar:{}}
   }
*/
```

**Reply by msageryd** [reply]
Somehow RHF must set the values deep in your object. I'm thinking that this might be a by-design-feature.

If you have registered your deep structure you essentially have the following fields:
- "isValid.true"
- "tag.foo"
- "tag.bar"

With this structure, you actually don't have a field named "isValid". You want to controll the deep structure by yourself, even whipe it out completely. I.e. only the first level in your object are actual fields. I think you should register your fields as just "isValid" and "tag". Then you have to take care of the deeper structure yourself.

**Reply by AveryTheDev** [reply]
> With this structure, you actually don't have a field named "isValid".

I do have a field explicitly called, 'isValid' and 'tag' in this case. They are already registered as such. They just have object values as their children. 

> You want to controll the deep structure by yourself, even whipe it out completely

I've attempted to handle the deeper structure with `setValue` however the values are not resetting. Is this methodology not valid?

```js
 const resetAllValues = () => {
    reset(defaultValues);
    setValue('tag', {}); //only the rest value is seen
   console.log(formData); 

/*
  {
    isValid: {true: undefined},
    tag: {foo: undefined, bar: undefined}
   }
*/
};
```

**Reply by msageryd** [reply]
In that case it seems really odd, and I'm also curious on how this works.
Can you share your code for registering those fields?

**Reply by AveryTheDev** [reply]
I'll have to make a sandbox without any company code in it for privacy concerns. This will take sometime since I'm about to go on my lunch break, but I'll get it done later today and post it as a reply here for visibility.

**Reply by msageryd** [reply]
I'm just interested in the actual `register()` call, or maybe `<Controller>` or `useControl()` depending on your choice.

**Comment by AveryTheDev** [comment, marked as answer]
Alright, I ended up solving it myself. 

The actual problem was a function I created to generate the default values according to what fields were being passed into it. There were some edge cases that I had to go back in and address, but once I addressed it everything was working as expected. 

Thank you both @msageryd and @Moshyfawn for helping me navigate this strange issue I was having!

**Comment by davidalvarezr** [comment]
Found a other way to solve this, calling `unregister("tag")` was the only way to remove nested "tag.foo" and "tag.bar":
https://react-hook-form.com/api/useform/unregister/
